client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -166,73 +167,159 @@ public Response performRequest(String method, String endpoint, Map<String, Strin
+                               ResponseListener responseListener, Header... headers) {
+        performRequest(method, endpoint, params, null, responseListener, headers);
+    }
+
+    /**
+     /**
+     * Sends a request to the elasticsearch cluster that the current client points to.
+     * Shortcut to {@link #performRequest(String, String, Map, HttpEntity, HttpAsyncResponseConsumer, ResponseListener, Header...)}
+     * but without an async response consumer, meaning that a {@link org.apache.http.nio.protocol.BasicAsyncResponseConsumer}
+     * will be created and used.
+     *
+     * @param method the http method
+     * @param endpoint the path of the request (without host and port)
+     * @param params the query_string parameters
+     * @param entity the body of the request, null if not applicable
+     * @param responseListener the {@link ResponseListener} to notify when the request is completed or fails
+     * @param headers the optional request headers
+     */
+    public void performRequest(String method, String endpoint, Map<String, String> params,
+                               HttpEntity entity, ResponseListener responseListener, Header... headers) {
+        HttpAsyncResponseConsumer<HttpResponse> responseConsumer = HttpAsyncMethods.createConsumer();
+        performRequest(method, endpoint, params, entity, responseConsumer, responseListener, headers);
+    }
+
+    /**
+     * Sends a request to the elasticsearch cluster that the current client points to. The request is executed asynchronously
+     * and the provided {@link ResponseListener} gets notified whenever it is completed or it fails.
+     * Selects a host out of the provided ones in a round-robin fashion. Failing hosts are marked dead and retried after a certain
+     * amount of time (minimum 1 minute, maximum 30 minutes), depending on how many times they previously failed (the more failures,
+     * the later they will be retried). In case of failures all of the alive nodes (or dead nodes that deserve a retry) are retried
+     * till one responds or none of them does, in which case an {@link IOException} will be thrown.
+     *
+     * @param method the http method
+     * @param endpoint the path of the request (without host and port)
+     * @param params the query_string parameters
+     * @param entity the body of the request, null if not applicable
+     * @param responseConsumer the {@link HttpAsyncResponseConsumer} callback
+     * @param responseListener the {@link ResponseListener} to notify when the request is completed or fails
+     * @param headers the optional request headers
+     */
+    public void performRequest(String method, String endpoint, Map<String, String> params,
+                               HttpEntity entity, HttpAsyncResponseConsumer<HttpResponse> responseConsumer,
+                               ResponseListener responseListener, Header... headers) {
```

Comment:
I'm torn here. I'd love to expose something that knows how to turn the response into the thing we *want* rather than an `HttpResponse`. It feels silly to buffer the whole thing on the heap and then spin it back into the response. But `HttpAsyncResponseConsumer` doesn't seem like it is the right thing to expose though because there are a lot of things that aren't needed for that conversion.